### ISSUE TITLE
How to create a new currency

### ISSUE THREAD

**Issue by Kyborg2011** [ton-blockchain/ton#206]
Hello, everyone!

As I see in documentation and white paper, there is an ability to create my own currency in TON, based on smart-contract, but I have not found any true guide about this option in any of current official documents, shared by TON creators. I have searched in: white paper, tblkch.pdf, tvm.pdf, fiftbase.pdf and in other text documents on a website (test.ton.org).

So, can you help me with it? **Also, is it real right now, in testnet2?** 

P.S.: I am asking this, as a  participant in Blockchain contest (stage 2) ;)

**Comment by ton-blockchain** [comment]
Yes. You have to send an internal message with a predefined amount of (test) grams to a special "minter" smartcontract, indicating an unused currency ID and the total amount of the new currency to be minted. You get the new tokens in an answer message.

What is your expected use of the new currency? If you just want to have some tokens distinct from test Grams, we can send you some.

**Comment by Kyborg2011** [comment]
@ton-blockchain Thank you!
I am thinking about an idea for a smart contract in 2nd stage of a competition, so I think about creating of a full Token by smart contract, like in Ethereum Blockchain - the most of smart contracts have it unique currency.

So, particular for a contest, is it needed - a unique currency, or not? If not I think I can make a work, based simply on a base currency (GRM).

I have a test grams ) Thank you! I have got it from an official Telegram account, when I was getting my prize In a 1st stage of a competition.

**Comment by Kyborg2011** [comment]
@ton-blockchain And another simple question: I need for my idea of a smart contract some functions for a pseudo random numbers generation. I have found in tvm.pdf that in Fift assembler there are some random generation functions, but I do not see there any real examples of them. Can you help me with that, too? )

**Comment by ton-blockchain** [comment]
What kind of random number generating functions would you like to have?
Random 256 bits? Random number 0 .. n-1 for any n?

**Comment by Kyborg2011** [comment]
@ton-blockchain Any of these variants is good for me, but right now for my idea - random generator of numbers 0 ... n-1 is enough for me,

**Comment by danog** [comment]
^, I also am currently in need of those pseudorandom primitives, 0 .. n-1 is also fine for me.

**Comment by ton-blockchain** [comment]
Well, the new `RAND` primitive might be what you want. It is documented in A.11.3 of [tvm.pdf](https://test.ton.org/tvm.pdf).

**Comment by Kyborg2011** [comment]
@ton-blockchain I have found it ) Thank you for so fast improvement!

**Comment by Skydev0h** [comment]
@ton-blockchain does such kind of minting work at all?
I checked throughtly in the C++ source code and can't find correct message layout.
Moreover, looking at a glance at -1:0 contract and looking at gen-zerostate I can see that -1:0 contract is just a very fat simple wallet.
Tried sending some kind of message carrying 32-bit key, 32-bit amount, but expectedly, the "minter" contract swallowed it.
Is it correct that currently minting is not supported?

**Comment by ton-blockchain** [comment]
Automatic minting is not enabled now, that's true. If you need some extra currencies (different from test grams) just to test something, we can send you some in the testnet. If you want to test the automatic creation of new currencies, you'll have to wait a bit.

**Comment by plexar88** [comment]
Can you send me please for testing purposes extra currency (id:241, amount:1000000000000) to wallet kQDEysEvW8fu9M9eyE7mjM-GCSGgbKA5XsVY5T43sTw7CLt1
